### ISSUE TITLE
istioctl: Support setting resource quotas on injected sidecars

### ISSUE THREAD

**Issue by taoyu27** [istio/istio#2864]
**env:**
```
centos7.2
istio version: 0.4.0
kubenetes version: v1.9.1
```
namespace has quota configuration

```
[root@master ~]# kubectl describe ns namespace-892ec724
Name:         namespace-892ec724
Labels:       <none>
Annotations:  <none>
Status:       Active

Resource Quotas
 Name:            namespace-892ec724-quota
 Resource         Used    Hard
 --------         ---     ---
 limits.cpu       1300m   10
 limits.memory    2560Mi  20Gi
 requests.cpu     1300m   10
 requests.memory  1792Mi  20Gi
```

**action:**
testing the function: istio automatic injection

**problem happen:**
E0126 07:48:29.380616       1 replica_set.go:424] Sync "namespace-a71c3b09/deployment-nginx-test-1-c696c6c86" failed with pods "deployment-nginx-test-1-c696c6c86-" is forbidden: exceeded quota: namespace-a71c3b09-quota, requested: limits.cpu=1100m,requests.cpu=1100m, us
ed: limits.cpu=9100m,requests.cpu=9, limited: limits.cpu=10,requests.cpu=10

**analysis**
check the deployment of nginx.
segment of yaml file:
```yaml
 initContainers:
      - args:
        - -p
        - "15001"
        - -u
        - "1337"
        image: istio/proxy_init:0.4.0
        imagePullPolicy: IfNotPresent
        name: istio-init
        resources: {}
        securityContext:
          capabilities:
            add:
            - NET_ADMIN
          privileged: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - emptyDir:
          medium: Memory
        name: istio-envoy
      - name: istio-certs
        secret:
          defaultMode: 420
          optional: true
          secretName: istio.default
```
we can fount quota configuration missing here when istio sidecar injected

According to segment of https://istio.io/docs/setup/kubernetes/sidecar-injection.html#configuration-options. I guess that there should be a flag to set the quota when sidecar template be loaded by initializers controller. who can tell me how to do ? thanks

**Comment by sak0** [comment]
same problem. 
there is a workaround: create limitrange for every namespaces. but we need set the istio-sidercar resource for defferent condition.

**Comment by mandarjog** [comment]
@ayj

**Comment by ayj** [comment]
This should be doable with the new templated configuration format. See https://github.com/istio/istio/blob/master/install/kubernetes/templates/istio-sidecar-injector-configmap-release.yaml.tmpl. You can add whatever you want to the template without needing to add new flags and rebuild binaries.

**Comment by taoyu27** [comment]
@ayj , I found that MutatingWebhookConfiguration was used in istio-0.4.0 later. if take new template effect, k8s need to support this new admission controller?

**Comment by ayj** [comment]
The mutating webhook is available starting in istio-0.5.0 and requires k8s 1.9. You'll need to use manual injection (`kube-inject`) on k8s 1.7 and 1.8. Both manual and automatic injection methods support the templated configuration.

**Comment by frankbu** [comment]
When this is implemented, we should also update the helloworld sample instructions to use it.
https://github.com/istio/istio/tree/master/samples/helloworld#start-the-services

**Comment by ayj** [comment]
@ldemailly, do you have any suggestions for reasonable default resource requests/limits for `istio-proxy`?

**Comment by ldemailly** [comment]
the unset default is probably fine for now, otherwise 0.9 vCPU, 250Mb ram ?

**Comment by taoyu27** [comment]
@ayj , I tried with istio-0.5.0, but one issue happen, could you please disabuse me of issue?
install steps:
1. kubectl apply -f istio.yaml
2. add admision controller to api-server with MutatingAdmissionWebhook
3. 
```
./install/kubernetes/webhook-create-signed-cert.sh \
    --service istio-sidecar-injector \
    --namespace istio-system \
    --secret sidecar-injector-certs
```
error happen this step:
```
ERROR: After approving csr istio-sidecar-injector.istio-system, the signed certificate did not appear on the resource. Giving up after 10 attempts.
See https://istio.io/docs/setup/kubernetes/sidecar-injection.html for more details on troubleshooting.
```

check the script, error generated in this line
```
serverCert=$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
```
execute this command on the k8s independently
```
[root@master kubernetes]# kubectl get csr
NAME                                  AGE       REQUESTOR          CONDITION
istio-sidecar-injector.istio-system   2m        system:unsecured   Approved
```
```
[root@master kubernetes]# kubectl get csr istio-sidecar-injector.istio-system -o yaml
apiVersion: certificates.k8s.io/v1beta1
kind: CertificateSigningRequest
metadata:
  creationTimestamp: 2018-02-11T08:55:16Z
  name: istio-sidecar-injector.istio-system
  resourceVersion: "416723"
  selfLink: /apis/certificates.k8s.io/v1beta1/certificatesigningrequests/istio-sidecar-injector.istio-system
  uid: 4767e6c2-0f09-11e8-8a7a-fa163e200e23
spec:
  groups:
  - system:masters
  - system:authenticated
  request: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURSBSRVFVRVNULS0tLS0KTUlJRE9EQ0NBaUFDQVFBd01qRXdNQzRHQTFVRUF4TW5hWE4wYVc4dGMybGtaV05oY2kxcGJtcGxZM1J2Y2k1cApjM1JwYnkxemVYTjBaVzB1YzNaak1JSUJJakFOQmdrcWhraUc5dzBCQVFFRkFBT0NBUThBTUlJQkNnS0NBUUVBCnJkQW9WZTZ3K1dwQkdmUHBtbUk5eWVHeXZacWN
0azU3WU1JOVB6aVlaNW9ZMG12TThpL05IUTRCdmlJN09NZDgKbDhaZUxDU1AwZUlNL0JZRUdLS04rY1YxRm4zWjBZUWdQYmdWREQ4MlFnQ0t0WFdqYmM1NFRnNmJwZEozZWVQQwpja2FsT05OTXF6MjZuT3BRYVlHT1NSTVkwR0dZWkNaUU0vTXJmR0NUemozWXY4NUdBeC91VzVqZ1FlR3FVZmNWClZ1VW04ZlgxNERqUEh1Qm8vUEpvRmZOSVZtbHo4RHkxY2E5SmhOTTNzY3JCdHpsV3BBWFlHWGk0bVNReFZOdkEKdk50ZXpjTjV3U0tHS1hRTURrOXRwd3FHTVVYLytKWEQxeUlTOWRvZjFJbXpMSjl3dEZ4SjhIWEF2Qm1oZTZFbApFbjByUU4rcWxmOW9xMEZNcHZUZnNRSURBUUFCb0lIQU1JRzlCZ2txaGtpRzl3MEJDUTR4Z2E4d2dhd3dDUVlEClZSMFRCQUl3QURBTEJnTlZIUThFQkFNQ0JlQXdFd1lEVlIwbEJBd3dDZ1lJS3dZQkJRVUhBd0V3ZlFZRFZSMFIKQkhZd2RJSVdhWE4wYVc4dGMybGtaV05oY2kxcGJtcGxZM1J2Y29JamFYTjBhVzh0YzJsa1pXTmhjaTFwYm1wbApZM1J2Y2k1cGMzUnBieTF6ZVhOMFpXMkNOV2x6ZEdsdkxYTnBaR1ZqWVhJdGFXNXFaV04wYjNJdWFYTjBhVzh0CmMzbHpkR1Z0TG5OMll5NWpiSFZ6ZEdWeUxteHZZMkZzTUEwR0NTcUdTSWIzRFFFQkJRVUFBNElCQVFBbm5vU04KazJvUE53SksvemlWSnlKNW5MeSs3eGNCbjdlOTl6NjVOS2hCdkloZFZuQTYwTUpQUVN2YjVsbGpmaE9PTWtBTwpoWjBEbVlFZS9wU0JrYkJZZnBFSyt6dkwzMjZKdFJtVytiZFhMcDR0TnNQeGxmdkxGdkhEVm1qZ1NBTFBSTFZmCjlmS01XL0FKY24veC9hOTBNMjAvYThWaDgzdlRPQWVGcjJxMmwyOFZUY3JmWFp0N0NiZFY2ZVJCQlBmRlhQZFkKcGNJTmR5bVdMUVlMME41V2d1TVRtTjI3ZS9mcTJSTS8wZ3dac05SdVBxc3pwcGpuWmhWNEdkWWJXSjlNZFc4Ugp4VWc2ZW9tR1pYYVo4bkUvRnVqL0ROY05NNExNc21hVjBwUGN4cXdvRTVBa25mMnZtVktCcFRjSDZxNjNtUyt1CkFPN3h5dEFRQnR2c2NZcXYKLS0tLS1FTkQgQ0VSVElGSUNBVEUgUkVRVUVTVC0tLS0tCg==  usages:
  - digital signature
  - key encipherment
  - server auth
  username: system:unsecured
status:
  conditions:
  - lastUpdateTime: 2018-02-11T08:55:16Z
    message: This CSR was approved by kubectl certificate approve.
    reason: KubectlApprove
    type: Approved
```

checked the results. no  "status.certificates".

how can i resolve this problem?  thanks a lot!

**Comment by taoyu27** [comment]
yet another:

error happen when create MutatingWebhookConfiguration

yaml file:
```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  name: istio-sidecar-injector
webhooks:
  - name: sidecar-injector.istio.io
    clientConfig:
      service:
        name: istio-sidecar-injector
        namespace: istio-system
        path: "/inject"
      caBundle: ${CA_BUNDLE}
    rules:
      - operations: [ "CREATE" ]
        apiGroups: [""]
        apiVersions: ["v1"]
        resources: ["pods"]
    namespaceSelector:
      matchLabels:
        istio-injection: enabled
```
result:
```
Error from server (BadRequest): error when creating "aa.yaml": MutatingWebhookConfiguration in version "v1beta1" cannot be handled as a MutatingWebhookConfiguration: v1beta1.MutatingWebhookConfiguration: Webhooks: []v1beta1.Webhook: v1beta1.Webhook: ClientConfig: v1beta
1.WebhookClientConfig: Service: CABundle: decode base64: illegal base64 data at input byte 0, parsing 859 ...A_BUNDLE}"... at {"apiVersion":"admissionregistration.k8s.io/v1beta1","kind":"MutatingWebhookConfiguration","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"admissionregistration.k8s.io/v1beta1\",\"kind\":\"MutatingWebhookConfiguration\",\"metadata\":{\"annotations\":{},\"name\":\"istio-sidecar-injector\",\"namespace\":\"\"},\"webhooks\":[{\"clientConfig\":{\"caBundle\":\"${CA_BUNDLE}\",\"service\":{\"name\":\"istio-sidecar-injector\",\"namespace\":\"istio-system\",\"path\":\"/inject\"}},\"name\":\"sidecar-injector.istio.io\",\"namespaceSelector\":{\"matchLabels\":{\"istio-injection\":\"enabled\"}},\"rules\":[{\"apiGroups\":[\"\"],\"apiVersions\":[\"v1\"],\"operations\":[\"CREATE\"],\"resources\":[\"pods\"]}]}]}\n"},"name":"istio-sidecar-injector","namespace":""},"webhooks":[{"clientConfig":{"caBundle":"${CA_BUNDLE}","service":{"name":"istio-sidecar-injector","namespace":"istio-system","path":"/inject"}},"name":"sidecar-injector.istio.io","namespaceSelector":{"matchLabels":{"istio-injection":"enabled"}},"rules":[{"apiGroups":[""],"apiVersions":["v1"],"operations":["CREATE"],"resources":["pods"]}]}]}
```
env:
k8s 1.9.1
api-server: 
```
- --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,ResourceQuota,Initializers,MutatingAdmissionWebhook
```

api supported:
```
{
  "paths": [
    "/api",
    "/api/v1",
    "/apis",
    "/apis/",
    "/apis/admissionregistration.k8s.io",
    "/apis/admissionregistration.k8s.io/v1alpha1",
    "/apis/admissionregistration.k8s.io/v1beta1",
```

**Comment by ayj** [comment]
Have you set the prerequisites correct? (see https://istio.io/docs/setup/kubernetes/sidecar-injection.html#prerequisites). The missing `status.certificate` suggests that the certificate controller wasn't started. If the problem persists, can you open a new github issue with the specifics?

**Comment by ayj** [comment]
> the unset default is probably fine for now, otherwise 0.9 vCPU, 250Mb ram ?

@costinm, @ldemailly  - what's the remaining specific feature request in this issue? User's can customize the the template and add their own sidecar resource limits right now. Do we need defaults (.9 vCPU _per sidecar is quite a lot) or yet-another-kube-inject flag?

**Comment by ldemailly** [comment]
@ayj maybe gen-deploy and/or helm values should cover changing limits but otherwise I think we can close this

**Comment by taoyu27** [comment]
@ayj , thanks for your response. 
but another error reported when create the sidercar-inject deployment.
```
Error: failed to create injection webhook tls: failed to find "CERTIFICATE" PEM block in certificate input after skipping PEM blocks of the following types: [CERTIFICATE REQUEST]
```
I don't know if relatived with issue #3123, so  I hava upgrade the istio to 0.5.1. but issue still exist. 
Do we still need open a new github issue trace this specifics?

**Comment by ayj** [comment]
https://github.com/istio/istio/pull/3123 wasn't included in 0.5.1 release. You'll need to grab the latest script from master or try 0.6 when its out. It's possibly related to https://github.com/istio/issues/issues/184. The webhook will fail to start if the provided cert/key are bad.

**Comment by taoyu27** [comment]
@ayj , this issue is solved in istio-0.6.0, i will close this item. thanks for your help on the way.

**Comment by yehaifeng** [comment]
I have the same thing. My kubernetes is v1.10.1. Istio is v0.7.1
And two args  for kube-controller-manager to solve:
--cluster-signing-cert-file=/etc/kubernetes/ssl/kube-ca.pem 
--cluster-signing-key-file=/etc/kubernetes/ssl/kube-ca-key.pem
